api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -809,6 +810,28 @@ message Cluster {
   // [#comment:TODO(incfly): add a detailed architecture doc on intended usage.]
   repeated TransportSocketMatch transport_socket_matches = 43;
 
+  // Optional matcher that selects a transport socket from
+  // :ref:`transport_socket_matches <envoy_v3_api_field_config.cluster.v3.Cluster.transport_socket_matches>`.
+  //
+  // This matcher uses the generic xDS matcher framework and may reference common
+  // network inputs such as destination IP, source IP, destination port, source port,
+  // Server Name Indication (SNI), Application-Layer Protocol Negotiation (ALPN), and
+  // dynamic metadata to select a named transport socket. This provides flexibility
+  // similar to the listener :ref:`filter_chain_matcher
+  // <envoy_v3_api_field_config.listener.v3.Listener.filter_chain_matcher>`.
+  //
+  // If this field is set, it takes precedence over legacy metadata-based selection
+  // performed by :ref:`transport_socket_matches
+  // <envoy_v3_api_field_config.cluster.v3.Cluster.transport_socket_matches>` alone.
+  // If the matcher does not yield a match, Envoy uses the default transport socket
+  // configured for the cluster.
+  //
+  // When using this field, each entry in
+  // :ref:`transport_socket_matches <envoy_v3_api_field_config.cluster.v3.Cluster.transport_socket_matches>`
+  // must have a unique ``name``. The matcher outcome is expected to reference one of
+  // these names.
+  xds.type.matcher.v3.Matcher transport_socket_matcher = 59;
```

Comment:
We have a similar problem and were thinking of improving the certificate selector - e.g. there is some function that computes a secret name based on the connection, and secrets are stored somewhere globally (e.g. the certificate provider framework is one such storage). Have you explored that?

Reply:
That's quite interesting and can solve some of our use-cases as well. I haven't explored that but could take a look. 

In this specific use-case I'm looking into, we also need to override the SNI in addition to the mTLS certificates and the only option I see is to have a matcher which can then pick a transport socket from the `matches` array based on the input like network-namespace, etc.

Reply:
I think this is probably valuable on its own, and looks almost ready. I can imagine the older matcher could be subsumed if we add some label matcher.